### PR TITLE
Add an error Callback to try with the '-r' argument when the callback…

### PR DESCRIPTION
**Pull request by ewongbb**
… fails

Current wip.  

TODO:
  i) need to figure out how to add a test for this.

### CODE REVIEW

File: master/buildbot/changes/hgpoller.py [hunk 1]
```diff
@@ -167,6 +167,17 @@ def _getChanges(self):
...
 
+        # some systems have old versions of Mercurial in /usr/bin
+        # (i.e. 1.4 in the case for CentOS systems) and
+        # -b option isn't available.   This should fallback gracefully
+        # to using -r.  Without this fallback, the addCallback() using
+        # the '-b' argument will fail without any indications in the
+        # log that something went wrong aside for an unhelpful traceback.
+        fallback_args = ['pull', '-r', self.branch, self.repourl]
+
+        d.addErrback(lambda _: utils.getProcessOutput(
```
Comment (tardyp) [inline master/buildbot/changes/hgpoller.py]: I would prefer this method to be rewritten in inlineCallback for readability.

File: master/buildbot/changes/hgpoller.py [hunk 2]
```diff
@@ -124,50 +125,66 @@ def process(output):
...
+
+            # This command always produces data on stderr, but we actually
+            # do not care about the stderr or stdout from this command.
+            # We set errortoo=True to avoid an errback from the deferred.
+            # The callback which will be added to this
+            # deferred will not use the response.
+            yield utils.getProcessOutput(
+                self.hgbin, args, path=self._absWorkdir(),
+                env=os.environ, errortoo=True)
+        except Exception:
```
Comment (tardyp) [inline master/buildbot/changes/hgpoller.py]: please put the correct exception there (not bare Exception)

File: master/buildbot/test/unit/test_changes_hgpoller.py [hunk 3]
```diff
@@ -140,9 +141,8 @@ def check_changes(_):
             self.assertEqual(change['branch'], 'default')
             self.assertEqual(change['comments'], 'This is rev 73591')
 
-        d.addCallback(check_changes)
-        d.addCallback(self.check_current_rev(73591))
-        return d
+        yield check_changes()
```
Comment (tardyp) [inline master/buildbot/test/unit/test_changes_hgpoller.py]: no need to keep the check_changes as a function

File: master/buildbot/test/unit/test_changes_hgpoller.py [hunk 4]
```diff
@@ -159,6 +159,9 @@ def test_poll_several_heads(self):
             gpo.Expect('hg', 'pull', '-b', 'default',
                        'ssh://example.com/foo/baz')
             .path('/some/dir'),
+            gpo.Expect('hg', 'pull', '-r', 'default',
```
Comment (tardyp) [inline master/buildbot/test/unit/test_changes_hgpoller.py]: in order for this to be expected, you need to have the previous command to fail.
So you need a separate test that checks the usecase of older hg

File: master/buildbot/changes/hgpoller.py [hunk 5]
```diff
@@ -124,50 +125,66 @@ def process(output):
...
 
+    @defer.inlineCallbacks
     def _initRepository(self):
         """Have mercurial init the workdir as a repository (hg init) if needed.
 
         hg init will also create all needed intermediate directories.
         """
         if self._isRepositoryReady():
-            return defer.succeed(None)
+            yield defer.succeed(None)
```
Comment (seankelly) [inline master/buildbot/changes/hgpoller.py]: Will this do an early return of `_initRepository`?

**Comment by ewongbb** [comment]
Can someone clarify as to what I'm doing wrong?  The coverage tests that fail with a "No Such file or directory" works locally.

**Pull request closed without merging**